### ISSUE TITLE
setup_call_cleanup/3: Cleanup is not reliably performed

### ISSUE THREAD

**Issue by triska** [infradig/trealla#22]
Currently, I get:

<pre>
<b>?- setup_call_cleanup(true, (false;true), Det = yes).</b>
true.
</pre>

Expected: `Det = yes`.

**Comment by infradig** [comment]
setup_call_cleanup/3 was removed to redo.

**Comment by infradig** [comment]
Trealla Prolog (c) Infradig 2020, v1.6.19
?- setup_call_cleanup(true, (false;true), Det = yes).
Det = yes.
?- setup_call_cleanup(true, (false;true), fail).
true.

**Comment by pmoura** [comment]
> Trealla Prolog (c) Infradig 2020, v1.6.19
> ?- setup_call_cleanup(true, (false;true), Det = yes).
> Det = yes.
> ?- setup_call_cleanup(true, (false;true), fail).
> true.

Run all tests in the Logtalk distribution. A single test fails now (compared with a good number of tests before this fix) due to the test goal succeeding non-deterministically. That single test failure doesn't seem to be related to the `setup_call_cleanup/3` implementation.

**Comment by infradig** [comment]
I see a lot of fails in the logtalk tests. Are we looking at the same thing?

**Comment by infradig** [comment]
To be specific, running in the tests/logtalk directory.

**Comment by pmoura** [comment]
> I see a lot of fails in the logtalk tests. Are we looking at the same thing?

No. I was referring to all the tests in the `logtalk3` directory. The results using 1.6.21 (thus, a few commits older) were:

```text
% 500 test sets: 396 completed, 79 skipped, 10 broken, 1 timeouts, 14 crashes
% 5504 tests: 66 skipped, 4595 passed, 843 failed
```

Several of these tests also check for determinism. But checking for determinism is no longer an issue, as I wrote. Btw, 843 test failures (plus the broken, timeout, and crashed test sets) sounds like a big number but keep in mind that there's a relatively small number of bugs causing most of these issues.
